File: app/src/test/java/bj/rxjavaexperimentation/utils/DateFormatterTest.java
package bj.rxjavaexperimentation.utils;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.text.ParseException;
import java.util.Date;

import bj.rxjavaexperimentation.utils.wrapper.DateUtilsWrapper;
import bj.rxjavaexperimentation.utils.wrapper.SimpleDateFormatWrapper;

import static junit.framework.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by Josh Laird on 17/04/2017.
 */
@RunWith(MockitoJUnitRunner.class)
public class DateFormatterTest
{
    private static final String isoDate = "2015-04-24T21:48:23-07:00";
    private DateFormatter dateFormatter;
    @Mock DateUtilsWrapper dateUtilsWrapper;
    @Mock SimpleDateFormatWrapper simpleDateFormatWrapper;

    @Before
    public void setup()
    {
        dateFormatter = new DateFormatter(dateUtilsWrapper, simpleDateFormatWrapper);
    }

    @Test
    public void formatValidString_returnsFormattedString() throws ParseException
    {
        when(simpleDateFormatWrapper.parse(isoDate)).thenReturn(new Date(1429937303000L));
        when(dateUtilsWrapper.getRelativeTimeSpanString(any(Long.class), any(Long.class), any(Long.class))).thenReturn("15 Apr 2015");

        assertEquals("15 Apr 2015", dateFormatter.formatIsoDate(isoDate));

        verify(simpleDateFormatWrapper, times(1)).parse(isoDate);
        verify(dateUtilsWrapper, times(1)).getRelativeTimeSpanString(any(Long.class), any(Long.class), any(Long.class));
    }

    @Test
    public void formatInvalidString_returnsEmptyString() throws ParseException
    {
        when(simpleDateFormatWrapper.parse("canBeAnything")).thenThrow(ParseException.class);

        assertEquals("", dateFormatter.formatIsoDate("canBeAnything"));

        verify(simpleDateFormatWrapper, times(1)).parse("canBeAnything");
    }
}

File: app/src/test/java/bj/vinylbrowser/utils/DateFormatterTest.kt
package bj.vinylbrowser.utils

import bj.vinylbrowser.wrappers.DateUtilsWrapper
import bj.vinylbrowser.wrappers.SimpleDateFormatWrapper
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.times
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import junit.framework.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.junit.MockitoJUnitRunner
import java.text.ParseException
import java.util.*

/**
 * Created by Josh Laird on 22/05/2017.
 */
@RunWith(MockitoJUnitRunner::class)
class DateFormatterTest {
    val isoDate = "2015-04-24T21:48:23-07:00"
    lateinit var dateFormatter: DateFormatter
    val dateUtilsWrapper: DateUtilsWrapper = mock()
    val simpleDateFormatWrapper: SimpleDateFormatWrapper = mock()

    @Before
    fun setup() {
        dateFormatter = DateFormatter(dateUtilsWrapper, simpleDateFormatWrapper)
    }

    @Test
    @Throws(ParseException::class)
    fun formatValidString_returnsFormattedString() {
        whenever(simpleDateFormatWrapper.parse(isoDate)).thenReturn(Date(1429937303000L))
        whenever(dateUtilsWrapper.getRelativeTimeSpanString(any<Long>(Long::class.java), any<Long>(Long::class.java), any<Long>(Long::class.java))).thenReturn("15 Apr 2015")

        assertEquals("15 Apr 2015", dateFormatter.formatIsoDate(isoDate))

        verify(simpleDateFormatWrapper, times(1)).parse(isoDate)
        verify(dateUtilsWrapper, times(1)).getRelativeTimeSpanString(any<Long>(Long::class.java), any<Long>(Long::class.java), any<Long>(Long::class.java))
    }

    @Test
    @Throws(ParseException::class)
    fun formatInvalidString_returnsEmptyString() {
        whenever(simpleDateFormatWrapper.parse("canBeAnything")).thenThrow(ParseException::class.java)

        assertEquals("", dateFormatter.formatIsoDate("canBeAnything"))

        verify(simpleDateFormatWrapper, times(1)).parse("canBeAnything")
    }
}